### DISCUSSION TITLE
Transaction primitive (cloud.AtomicFunction)

### DISCUSSION THREAD

**Discussion by Chriscbr** [winglang/wing#6869] [Ideas]
### Feature Spec

As a user, I would like a way to perform multiple "transactions" in an atomic fashion, so that if any transaction fails, then they are all rolled back.

Pseudocode:

```js
bring rand;

let bucket = new cloud.Bucket();

let fn = new cloud.AtomicFunction(inflight () => {
  bucket.put("key1", "value1");
  if rand.random() > 0.5 {
    bucket.put("key2", "value2");
  } else {
    throw("network error");
  }
});

let reset = inflight () => {
  bucket.tryDelete("key1");
  bucket.tryDelete("key2");
};

test "atomic operations are always atomic" {
  for _ in 0..10 {
    fn.invoke();
    // either both objects exist, or neither object exists
    assert(bucket.exists("key1") == bucket.exists("key2");
    reset();
  }
}
```

### Use Cases

Performing safe, distributed transactions on the cloud when your state is located in multiple places or requires multiple operations. For example, one may want to model that a withdrawal from bank account A succeeds if and only if a deposit to bank account B succeeds.

### Implementation Notes

The implementation would heavily depend on which cloud target you are using. For example, on Azure you might be able to use Azure Durable functions, while on other clouds like AWS you might have to use AWS Step Functions or some other combination of services:

<img src="https://docs.aws.amazon.com/images/prescriptive-guidance/latest/patterns/images/pattern-img/fec0789c-d9b1-4d80-b179-dd9a7ecbec07/images/daad3e8e-6e6b-41c2-95c1-ca79d53ead64.png" width="600px">

### Component

Language Design, SDK

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

**Comment by Chriscbr** [comment]
Actually I realize there's some more nuance to the use case I wrote up -- I wrote above that what you'd want is "if any transaction fails, then they are all rolled back", but I think in practice what you want to do if a transaction fails might depend on the situation. If it's a database operation, sure you can rollback (or if it's a change to an S3 object, you can revert the object to a previous version if you have bucket versioning enabled). But for other kinds of operations, you might want to just fail, or add a retry policy, etc.

[a comment by an automated account is not shown]

**Comment by ekeren** [comment]
Keep

[a comment by an automated account is not shown]

**Comment by Chriscbr** [comment]
Converting this feature request into a discussion
